src/runtime/topology.rs

```diff
@@ -118,11 +119,18 @@ impl Topology {
     }
 
     /// Adds a [Block] to the [Topology] returning the `id` of the [Block] in the [Topology].
-    pub fn add_block(&mut self, mut block: Block) -> usize {
-        let block_name = block.type_name();
-        let block_id = self.blocks.vacant_key();
-        block.set_instance_name(format!("{}-{}", block_name, block_id));
-        self.blocks.insert(Some(block))
+    pub fn add_block(&mut self, mut block: Block) -> Result<usize> {
+        if let Some(name) = block.instance_name() {
+            if self.block_id(name).is_some() {
+                Err(Error::DuplicateBlockName(name.to_string()))?;
```

Comment:
Didn't know that this works. Can we do `return Err(...)`?

Reply:
It works, but that combination of using `anyhow::Result` and `?` results in an implicit conversion. Another reason to use `return` instead.